File: etl/lib/etl/docx_document.rb
require "zip"
require "nokogiri"

module Etl
  # Low level reader for the .docx (OOXML) format.
  #
  # Walks word/document.xml and returns an ordered list of "blocks", where
  # each block is either a paragraph of text or a table of rows/cells. This
  # keeps the reader generic and independent of the specific report layout;
  # the report-specific field extraction lives in ReportExtractor.
  class DocxDocument
    W_NS = "http://schemas.openxmlformats.org/wordprocessingml/2006/main".freeze

    Block = Struct.new(:type, :text, :rows, keyword_init: true)

    attr_reader :blocks

    def initialize(path)
      @path = path
      @blocks = read_blocks
    end

    # Full document text, one paragraph per line, tables rendered as
    # pipe-separated rows. Useful for narrative/keyword searches.
    def full_text
      blocks.map do |block|
        if block.type == :paragraph
          block.text
        else
          block.rows.map { |row| row.join(" | ") }.join("\n")
        end
      end.join("\n")
    end

    def paragraphs
      blocks.select { |b| b.type == :paragraph }.map(&:text)
    end

    def tables
      blocks.select { |b| b.type == :table }.map(&:rows)
    end

    private

    def read_blocks
      xml = Zip::File.open(@path) { |zip| zip.read("word/document.xml") }
      doc = Nokogiri::XML(xml)
      body = doc.at_xpath("//w:body", "w" => W_NS)
      return [] unless body

      body.children.filter_map { |node| parse_node(node) }
    end

    def parse_node(node)
      case node.name
      when "p"
        Block.new(type: :paragraph, text: paragraph_text(node))
      when "tbl"
        Block.new(type: :table, rows: table_rows(node))
      end
    end

    def paragraph_text(p_node)
      text = +""
      p_node.xpath(".//w:t | .//w:tab | .//w:br", "w" => W_NS).each do |node|
        case node.name
        when "t"  then text << node.text
        when "tab" then text << "\t"
        when "br" then text << "\n"
        end
      end
      text.strip
    end

    def table_rows(tbl_node)
      tbl_node.xpath("./w:tr", "w" => W_NS).map do |tr|
        tr.xpath("./w:tc", "w" => W_NS).map do |tc|
          tc.xpath(".//w:p", "w" => W_NS).map { |p| paragraph_text(p) }.reject(&:empty?).join(" ")
        end
      end
    end
  end
end

File: desktop/etl/docxDocument.js
"use strict";

const fs = require("node:fs");
const JSZip = require("jszip");
const { DOMParser } = require("@xmldom/xmldom");

const W_NS = "http://schemas.openxmlformats.org/wordprocessingml/2006/main";

function isW(node, localName) {
  return !!node && node.nodeType === 1 && node.localName === localName && node.namespaceURI === W_NS;
}

function walkElements(node, visit) {
  const children = node.childNodes;
  if (!children) return;
  for (let i = 0; i < children.length; i += 1) {
    const child = children.item(i);
    if (child.nodeType === 1) {
      visit(child);
      walkElements(child, visit);
    }
  }
}

function directChildren(node, localName) {
  const result = [];
  const children = node.childNodes;
  if (!children) return result;
  for (let i = 0; i < children.length; i += 1) {
    const child = children.item(i);
    if (isW(child, localName)) result.push(child);
  }
  return result;
}

function paragraphText(pNode) {
  let text = "";
  walkElements(pNode, (node) => {
    if (isW(node, "t")) text += node.textContent || "";
    else if (isW(node, "tab")) text += "\t";
    else if (isW(node, "br")) text += "\n";
  });
  return text.trim();
}

function tableRows(tblNode) {
  return directChildren(tblNode, "tr").map((tr) =>
    directChildren(tr, "tc").map((tc) => {
      const cellParagraphs = [];
      walkElements(tc, (node) => {
        if (isW(node, "p")) cellParagraphs.push(paragraphText(node));
      });
      return cellParagraphs.filter((text) => text !== "").join(" ");
    })
  );
}

function findBody(doc) {
  const bodies = doc.getElementsByTagNameNS(W_NS, "body");
  return bodies.length > 0 ? bodies.item(0) : null;
}

function readBlocks(doc) {
  const body = findBody(doc);
  if (!body) return [];

  const blocks = [];
  const children = body.childNodes;
  for (let i = 0; i < children.length; i += 1) {
    const child = children.item(i);
    if (isW(child, "p")) {
      blocks.push({ type: "paragraph", text: paragraphText(child) });
    } else if (isW(child, "tbl")) {
      blocks.push({ type: "table", rows: tableRows(child) });
    }
  }
  return blocks;
}

class DocxDocument {
  static async load(filePath) {
    const buffer = fs.readFileSync(filePath);
    const zip = await JSZip.loadAsync(buffer);
    const entry = zip.file("word/document.xml");
    if (!entry) throw new Error(`word/document.xml not found in ${filePath} - is this a valid .docx?`);

    const xml = await entry.async("string");
    const doc = new DOMParser().parseFromString(xml, "text/xml");
    return new DocxDocument(readBlocks(doc));
  }

  constructor(blocks) {
    this.blocks = blocks;
  }

  get fullText() {
    return this.blocks
      .map((block) => (block.type === "paragraph" ? block.text : block.rows.map((row) => row.join(" | ")).join("\n")))
      .join("\n");
  }

  get paragraphs() {
    return this.blocks.filter((b) => b.type === "paragraph").map((b) => b.text);
  }

  get tables() {
    return this.blocks.filter((b) => b.type === "table").map((b) => b.rows);
  }
}

module.exports = { DocxDocument };
